### ISSUE TITLE
`type` class cannot be used as a generic

### ISSUE THREAD

**Issue by dzshn** [python/mypy#11424]
(moved from python/typeshed#6218)

It seems as of python 3.9, `type[...]` has a special case for returning `GenericAlias` while not having a `__class_getitem__`, this causes issues when using it in favour of `typing.Type`:

```sh
$ python -c 'type[int]'  # Works
$ mypy -c 'type[int]'
<string>:1: error: Value of type "Type[type]" is not indexable
Found 1 error in 1 file (checked 1 source file)
```

I stumbled upon this issue with `python 3.9.7`/`mypy 0.910` but it is still there as of `mypy 0.920+dev.8b9ade2`

**Comment by hauntsaninja** [comment]
Oh, interesting. The following do work on master:
```
λ mypy -c 'z: type[int]' --python-version 3.9 
Success: no issues found in 1 source file
λ mypy -c 'z = type[int]' --python-version 3.9
Success: no issues found in 1 source file
```

Clearly a bug, curious if you have a more complex example that fails on master as well?

**Comment by dzshn** [comment]
Hmm, this is weird:
```sh
$ python -m mypy --version
mypy 0.920+dev.0ec868fe778c57b52f6df54f54eb19616e1d6ba2
$ python -m mypy -c 'type[int]' --python-version 3.9 
<string>:1: error: Value of type "Type[type]" is not indexable
Found 1 error in 1 file (checked 1 source file)

$ python -m mypy -c 'foo: type[int]' --python-version 3.9 
Success: no issues found in 1 source file

$ python -m mypy -c 'reveal_type(type[int]); foo: type[int]; reveal_type(foo)' --python-version 3.9 
<string>:1: error: Value of type "Type[type]" is not indexable
<string>:1: note: Revealed type is "Any"
<string>:1: note: Revealed type is "Type[builtins.int]"
Found 1 error in 1 file (checked 1 source file)
```

I also can't seem to reproduce the issue I was having, which was using something like `def foo(cls: type[T]) -> T` (or, I forgot some detail).. I guess this is low priority then?

**Comment by AlexWaygood** [comment]
No longer reproducing on 0.942

**Comment by mthuurne** [comment]
I can still reproduce this:
```
$ mypy --python-version 3.10 -c 'type[int]'
<string>:1: error: Value of type "Type[type]" is not indexable  [index]
Found 1 error in 1 file (checked 1 source file)
$ mypy --version
mypy 0.950 (compiled: yes)
```
It might be a duplicate of #12392 though.

**Comment by AlexWaygood** [comment]
> I can still reproduce this:
> 
> ```
> $ mypy --python-version 3.10 -c 'type[int]'
> <string>:1: error: Value of type "Type[type]" is not indexable  [index]
> Found 1 error in 1 file (checked 1 source file)
> $ mypy --version
> mypy 0.950 (compiled: yes)
> ```
> 
> And it also reproduces for me on 0.942.
> 
> It might be a duplicate of #12392 though.

Thanks, you're right -- not sure why I closed this. As you say, though, there's heavy overlap with #12392, which has more practical use cases in it. So I don't think there's a need to reopen, personally.

**Comment by mthuurne** [comment]
I can still reproduce the issue even after the #12392 fix: it seems that `type[X]` is handled when it occurs in an annotation or a potential alias, but not when it's used as any other kind of expression:

```py
def f() -> type:
    a: type[int]  # fine
    A = type[int]  # fine
    return type[int]  # error: Value of type "type[type]" is not indexable  [index]
```

I'm using mypy 1.4.0 on Python 3.10.6.

Could this issue be re-opened?

**Comment by AlexWaygood** [comment]
> I can still reproduce the issue even after the #12392 fix: it seems that `type[X]` is handled when it occurs in an annotation or a potential alias, but not when it's used as any other kind of expression:
> 
> ```python
> def f() -> type:
>     a: type[int]  # fine
>     A = type[int]  # fine
>     return type[int]  # error: Value of type "type[type]" is not indexable  [index]
> ```
> 
> I'm using mypy 1.4.0 on Python 3.10.6.
> 
> Could this issue be re-opened?

This is basically covered by #13026 I think

**Comment by mthuurne** [comment]
Agreed, no need to re-open then.
